### PR TITLE
Merge ported components of the OpenJ9 shared class cache into OMR.

### PR DESCRIPTION
**Pull request by mthom**
This pull request includes needed functions missing from the OMR port library,
such as the sysvipc routines.

Issue: #4610

Signed-off-by: Mark Thom <mark.thom@unb.ca>

### CODE REVIEW

**Comment by 0xdaryl** [comment]
Please investigate the Windows CI failures as a starting point.

**Comment by 0xdaryl** [comment]
I'd like to request help reviewing this PR and the design from any or all OpenJ9 shared classes and port library subject matter experts.  @hangshao0 @rwy0717 @youngar @DanHeidinga @mpirvu @dsouzai @mstoodle

**Comment by 0xdaryl** [comment]
This also seems like a good topic for the OMR Architecture meeting on Jan. 16, 2020.

File: include_core/omrport.h [hunk 1]
```diff
@@ -1910,6 +2054,94 @@ typedef struct OMRPortLibrary {
 	uintptr_t (*mmap_get_region_granularity)(struct OMRPortLibrary *portLibrary, void *address) ;
 	/** see @ref omrmmap.c::omrmmap_dont_need "omrmmap_dont_need"*/
 	void (*mmap_dont_need)(struct OMRPortLibrary *portLibrary, const void *startAddress, size_t length) ;
+	/** see @ref j9shsem.c::j9shsem_params_init "j9shsem_params_init"*/
```
Comment (DanHeidinga) [inline include_core/omrport.h]:
```suggestion
	/** see @ref omrshsem.c::omrshsem_params_init "omrshsem_params_init"*/
```

File: include_core/shchelp.h [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+#define OMRSH_MODLEVEL_JAVA5  1
+#define OMRSH_MODLEVEL_JAVA6  2
+#define OMRSH_MODLEVEL_JAVA7  3
+#define OMRSH_MODLEVEL_JAVA8  4
+#define OMRSH_MODLEVEL_JAVA9  5
+/*
+ * From Java 10, modLevel equals to java version number. But there might be Java 10 shared cache that has modLevel 6
+ * created before this change. Define OMRSH_MODLEVEL_JAVA10 to handle them
+ */
+#define OMRSH_MODLEVEL_JAVA10  6
```
Comment (DanHeidinga) [inline include_core/shchelp.h]: Lots of Java specific constants here - OMRSH_MODLEVEL_**JAVA**5

File: include_core/shchelp.h [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+#define OMRPORT_SHR_CACHE_TYPE_NONPERSISTENT  2
+#define OMRPORT_SHR_CACHE_TYPE_VMEM  3
+#define OMRPORT_SHR_CACHE_TYPE_CROSSGUEST  4
+#define OMRPORT_SHR_CACHE_TYPE_SNAPSHOT  5
+
+uintptr_t
+getValuesFromShcFilePrefix(struct OMRPortLibrary* portLibrary, const char* filename, struct J9PortShcVersion* versionData);
+
+//uint32_t
+//getShcModlevelForJCL(uintptr_t j2seVersion);
```
Comment (DanHeidinga) [inline include_core/shchelp.h]: Delete commented code

File: include_core/shchelp.h [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+uint32_t
+getJCLForShcModlevel(uintptr_t modlevel);
```
Comment (DanHeidinga) [inline include_core/shchelp.h]: JCL = Java Class Library - intentional here?

File: port/CMakeLists.txt [hunk 5]
```diff
@@ -169,6 +169,8 @@ list(APPEND OBJECTS
 	omrmemcategories.c
 	omrport.c
 	omrmmap.c
+        omrargscan.c
+        omrshchelp.c
```
Comment (DanHeidinga) [inline port/CMakeLists.txt]: tabs vs space

File: port/common/omrargscan.c [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+ *******************************************************************************/
+
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <math.h>
+#include "omr.h"
+#include "omrargscan.h"
+#include "omrport.h"
+//#include "jvminit.h"
```
Comment (DanHeidinga) [inline port/common/omrargscan.c]: delete

File: port/common/omrshchelp.c [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+#include <string.h>
+//#include "j9cfg.h"
+#include "shchelp.h"
+#include "omrutil.h"
+#include "omrargscan.h"
+#include "omrport.h"
+//#include "j2sever.h"
```
Comment (DanHeidinga) [inline port/common/omrshchelp.c]: delete commented includes

File: port/common/omrshchelp.c [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+//		break;
+//	case J2SE_19 :
+//		modLevel = OMRSH_MODLEVEL_JAVA9;
+//		break;
+//	default: 
+//		modLevel = JAVA_SPEC_VERSION_FROM_J2SE(j2seVersion);
+//		break;
+//	}
+//	return modLevel;
+//}
```
Comment (DanHeidinga) [inline port/common/omrshchelp.c]: delete

File: port/common/omrshchelp.c [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+
+/**
+ * Return the java spec version for a given modification level
+ * @param [in] modlevel  The modification level
+ * 
+ * @return the j2se version or 0 if one cannot be found
+ */
+uint32_t
+getJCLForShcModlevel(uintptr_t modlevel)
```
Comment (DanHeidinga) [inline port/common/omrshchelp.c]: delete?

File: port/common/omrshchelp.c [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+
+// 	if ((versionData.esVersionMajor == EsVersionMajor) &&
+//			(versionData.esVersionMinor == EsVersionMinor) &&
+//			(jclLevel == javaVersion) &&
+//			(versionData.addrmode == OMRSH_ADDRMODE) &&
+//			(versionData.feature == feature)) {
+//		return 1;
+//	}
+	
+//	return 0;
```
Comment (DanHeidinga) [inline port/common/omrshchelp.c]: delete commented code

File: port/common/omrshchelp.c [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+
+/**
+ * Get the string for a given j2se version
+ * 
+ * @param [in] portlib  A port library
+ * @param [in] modlevel  The modification level
+ * @param [out] buffer  The buffer to copy the string into
+ */
+void
+getStringForShcModlevel(OMRPortLibrary* portlib, uint32_t modlevel, char* buffer, uint32_t buffSize)
```
Comment (DanHeidinga) [inline port/common/omrshchelp.c]: This looks like another java-ism

File: port/common/omrshchelp.c [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+	J9PortShcVersion versionData;
+	intptr_t nameToTestLen = 0;
+	intptr_t expectedVersionLen = ((expectCacheType == OMRPORT_SHR_CACHE_TYPE_PERSISTENT) || (OMRPORT_SHR_CACHE_TYPE_SNAPSHOT == expectCacheType)) ? OMRSH_VERSION_STRING_LEN + 1 : OMRSH_VERSION_STRING_LEN;
+	uintptr_t generation = getGenerationFromName(nameToTest);
+	intptr_t modLevel = 0;
+
+	/*
+	 * cache names generated by earlier JVMs (generation <=29) don't have feature prefix char 'F' and the value
+	 * adjust versionLen to parse cacheNameWithVGen generated by earlier JVMs according to generation number
+	 */
```
Comment (DanHeidinga) [inline port/common/omrshchelp.c]: java ism?

**Comment by hangshao0** [comment]
I agree that several smaller PRs makes it easier to review.

**Comment by mthom** [comment]
In response to the recent comments here, I've broken this PR up into these components:

https://github.com/eclipse/omr/pull/4756
https://github.com/eclipse/omr/pull/4755
https://github.com/eclipse/omr/pull/4754
https://github.com/eclipse/omr/pull/4753
https://github.com/eclipse/omr/pull/4752
https://github.com/eclipse/omr/pull/4751
https://github.com/eclipse/omr/pull/4750
https://github.com/eclipse/omr/pull/4749

**Pull request closed without merging**